include/seqan3/io/stream/parse_condition.hpp

```diff
@@ -0,0 +1,470 @@
+    //!\}
+
+    /*!\name Logical operators
+     * \brief Adds logical operators to allow logical disjunction, conjunction and negation on parse conditions.
+     * \{
+     */
+    //!\brief Combines the result of two seqan3::detail::parse_condition via logical disjunction.
+    template <parse_condition_concept rhs_condition_t>
+    constexpr auto operator||(parse_condition<rhs_condition_t> const &) const
+    {
+        // using new_condition_t = decltype(or_fn{cond, rhs.cond});
+        return or_fn<condition_t, rhs_condition_t>{};
+    }
+
+    //!\brief Combines the result of two seqan3::detail::parse_condition via logical conjunction.
+    template <parse_condition_concept rhs_condition_t>
+    constexpr auto operator&&(parse_condition<rhs_condition_t> const &) const
+    {
+        // using new_condition_t = decltype(and_fn<condition_t, rhs_condition_t>);
+        return and_fn<condition_t, rhs_condition_t>{};
+    }
+
+    //!\brief Negates the result of a seqan3::detail::parse_condition
+    constexpr auto operator!() const
+    {
+        // using new_condition_t = decltype(not_fn{cond});
+        return not_fn<condition_t>{};
+    }
+    //!\}
+
+    /*!\name Function call operator
+     * \{
+     */
+    //!\brief Invokes the condition on `val`.
+    template <alphabet_concept value_t>
+    bool operator()(value_t && val) const
+    {
+        return std::invoke(condition_t(), std::forward<value_t>(val));
+    }
+    //!\}
+
+    /*!\name Output functions
+     * \{
+     */
+    //!\brief Returns the message representing this condition as std::string.
+    //!\cond
+    template <typename t = condition_t>
+        requires is_parse_condition<t>
+    //!\endcond
+    std::string message() const
+    {
+        return condition_t::msg.string();
+    }
+    //!\}
+};
+
+// ----------------------------------------------------------------------------
+// or_fn
+// ----------------------------------------------------------------------------
+
+/*!\brief Logical disjunction operator for parse conditions.
+ * \implements seqan3::detail::parse_condition_concept
+ * \tparam condition_ts Template parameter pack over all parse condition types. Must contain at least 2 template parameters.
+ *                      Must satisfy the seqan3::detail::parse_condition_concept.
+ * \ingroup io
+ */
+template <parse_condition_concept... condition_ts>
+//\cond
+    requires sizeof...(condition_ts) >= 2
+//\endcond
+struct or_fn : public parse_condition<or_fn<condition_ts...>>
+{
+    //!\brief The message representing the disjunction of the associated conditions.
+    static constexpr auto msg = detail::condition_message_v<'|', condition_ts...>;
+
+    //!\brief Calls each parse condition with `val` and returns the logical disjunction of their results.
+    template <typename value_t>
+    inline bool operator()(value_t && val) const
+    {
+        return (... || std::invoke(condition_ts(), std::forward<value_t>(val)));
+    }
+};
```

Comment:
Should we maybe add overloads that optimize certain conditions combinations so that they aren't checked twice, e.g. same condition checked twice or overlapping ranges of.values?

Reply:
It is a nice thought, but I would rather not preoptimize stuff. From our current use cases it doesn‘t seem to be a real issue.